### ISSUE TITLE
Avoid the infinite loop when access_denied

### ISSUE THREAD

**Issue by CadiChris** [salvoravida/react-adal#33]
Hi there,

I'm facing the _infinite loop_ issue that are reported on adal repo (for instance [#635](https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/635) & [#463](https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/463)), but with React. 
Adal repo is providing guidelines on how to avoid it, but these guidelines [are tied to AngularJS](https://github.com/AzureAD/azure-activedirectory-library-for-js/wiki/FAQs#q2-my-app-gets-into-an-infinite-loop-sometimes-leading-to-digest-iterations-error).

**What I'd like to implement**:  if a User has no access to my app (because he's not allowed to in Azure AD), then I want to show him an `Access denied` page

**Where I am so far**
1. User browse my app to `http://myapp/` 
1. This triggers the Azure authentication windows
1. User authenticates
1. AzureAD understands my User has no rights, so it redirects to `http://myapp/#error=access_denied`
1. At this point, `react-adal` enters [this `if`](https://github.com/salvoravida/react-adal/blob/b1862045d491d8779877f54224b55ba9ba97c02b/src/react-adal.js#L66), so it triggers `authContext.login();`, which brings me back to step 2.

And there goes the infinite loop.

To put it in other words : I'm never able to reach the `<ErrorPage />` of
```js
const MyProtectedPage = withAdalLoginApi(
  MyPage, 
  () => <Loading />, 
  (error) => <ErrorPage error={error}/>
);
```

Have you already faced this `access_denied` loop problem ? Am I missing something obvious ?

**Comment by CadiChris** [comment]
I've also tried to set `redirectUri` to a URI that does **not** require authentication: I still get the infinite loop.

For instance

```js
// adal.js
const adalContext = new AuthenticationContext({
  //...
  redirectUri: "http://myapp/authentication-result",
});

// In index.js 
// I've set DO_NOT_LOGIN to true

// App.js
<Switch>
    <Route
          exact
          path="/" // When browsing this, I get the Azure authentication window
          render={() => <MyProtectedPage />} 
        />

    <Route
          // When browsing this, I don't get authentication window, as expected
          path="/authentication-result" 
          render={() => <h1>Authentication</h1>}  
        />

    <Route path="*">
          <Redirect to="/" />
    </Route>
</Switch>
```

With this setup, Azure properly redirects to `http://myapp/authentication-result/#error=access_denied`, but again a refresh occurs and brings me back to `http://myapp/`, that asks for authentication...and there goes infinite loop

**Comment by salvoravida** [comment]
if (msg === 'login required') {
              authContext.login();
            } else {
              this.setState({ error });
            }

if the msg is access_denied then the error component should render.

could you please check the  console.log(error); that you receive?

**Comment by CadiChris** [comment]
> could you please check the console.log(error); that you receive?

The error I have is `{message: "User login is required", msg: "login required"}`.

I've tried to understand precisely what's happening, step by step.
The following message is _kind of_ intended as [rubber duck debugging](https://en.wikipedia.org/wiki/Rubber_duck_debugging)
Here is the result of the investigation :

- I browse to `http://myapp/`
- `runWithAdal` is executed
  - it calls `authContext.handleWindowCallback();`, which at this point **doesn't do anything**, since there's no hash in the URL
   - then, since `doNotLogin` is set to `true`, `runWithAdal` renders `app()`

At this point my `<ProtectedPage />` is mounted, since it's mapped to the `myapp/` route.
- `componentWillMount()` of `withAdalLogin` is executed
   - it calls `authContext.aquireToken()`. `acquireToken` fails, since I'm not logged in, so it calls `callback("login required")`
   - back to `componentWillMount()`, we're now in the [`catch` block](https://github.com/salvoravida/react-adal/blob/master/src/react-adal.js#L63) that will lead to [`authContext.login();`](https://github.com/salvoravida/react-adal/blob/b1862045d491d8779877f54224b55ba9ba97c02b/src/react-adal.js#L67)

At this point, the Microsoft invite is displayed, so that I can log.
I log, with an **unauthorized account**, I'm then redirected to `http://myapp/#error=access_denied`
- `runWithAdal` is executed
  - it calls `authContext.handleWindowCallback();`
    - as opposed to the first time, this time there's an actuel hash in the URL, so `handleWindowCallback()` will
      - save the `access_denied` error in the local storage
      - [set `window.location.href` to `/`](https://github.com/salvoravida/react-adal/blob/master/src/adal.js#L1365), since `navigateToLoginRequestUrl` is set to `true` (the default value)
  - back to `runWithAdal`: it won't do anything else since [we're in the callback case](https://github.com/salvoravida/react-adal/blob/master/src/react-adal.js#L23)

So now...what happens is that since `window.location.href` has been set to `/`, `<ProtectedPage />` will be mounted, which will call `authContext.aquireToken()`, which will lead to the redirection to `http://myapp/#error=access_denied` etc. etc. there goes the loop.

I don't really know where I could or should change something...
Event if I set `navigateToLoginRequestUrl` to `false`, then adal will [simply remove the hash](https://github.com/salvoravida/react-adal/blob/b1862045d491d8779877f54224b55ba9ba97c02b/src/adal.js#L1366), but `<ProtectedPage />` will still be mounted, which will trigger infinite loop.

Again, I'm still in the process of trying to figure out a correct way of approaching this...if you have any hint, it'd be much appreciated.

**Comment by CadiChris** [comment]
I've come up with a working solution. 
I'm not entirely sure of how it fits into the whole adal process though.

What I thought of is changing `withAdalLogin` so that it does not `authContext.login()` if there's an existing `loggingError`.

With some code, related to [withAdalLogin](https://github.com/salvoravida/react-adal/blob/b1862045d491d8779877f54224b55ba9ba97c02b/src/react-adal.js#L60)
```js
    componentWillMount = () => {
      adalGetToken(authContext, resourceId)
        .then(() => this.setState({ logged: true }))
        .catch((error) => {
           const { msg } = error;
           console.log(error);

           const loginError = authContext.getLoginError();
           const loginWasTriedButFailed = loginError !== undefined && loginError !== ""
           if (loginWasTriedButFailed) {
              this.setState({error: loginError})
            }
            else if (msg === 'login required') {
              authContext.login();
            } else {
              this.setState({ error });
            }
          });
      };

```

What do you think @salvoravida ? If you think the idea is good, I'd be delighted to send a PR with a slightly more polished code.

**Comment by salvoravida** [comment]
@cadichris yes i agree with your workaround.
the problem is with official adal.js

AuthenticationContext.prototype.acquireToken 

that call -> callback('User login is required', null, 'login required');  if (!this.user)
without check that previous login failed.

**Comment by CadiChris** [comment]
Thanks @salvoravida. 
- Do you think the workaround is worth a PR here ?
- I'll open an issue on adal repo, to discuss this design, and see what their opinion is. _[EDIT]: it's the issue below._

**Comment by panuhorsmalahti** [comment]
My workaround (without forking react-adal or adal) is like this:

```
const loginError = authContext.getLoginError();
const loginWasTriedButFailed = loginError !== undefined && loginError !== null && loginError !== "";
const redirectIfNoUser = !loginWasTriedButFailed;

runWithAdal(
    authContext,
    () => {
        if (loginWasTriedButFailed) {
            history.push("/access-denied");
        }
        ReactDOM.render(...);
    },
    !redirectIfNoUser
);
```

With this:
* If user is not logged in at all, page is redirected to Azure login
* If user is logged in to Azure but doesn't have access to this app, access denied page is displayed (with log out button)
* If user is logged in and has access to the app, normal application is rendered

**Comment by zsid** [comment]
I am also having the same problem. @panuhorsmalahti when you redirect to `/access-denied`, are you also having protected routes with `react-adal` so that when it lands on `/access-denied` it doesn't redirect back to the Azure login page?

**Comment by panuhorsmalahti** [comment]
The `history.push` is a page change inside a single-page app, so it's not going through the logic again - therefore no redirect back to Azure.

**Comment by salvoravida** [comment]
should be fixed in next release.

**Comment by salvoravida** [comment]
@cadichris thanks for suggestion. it should be fixed on v0.5.0 

let me know your feedback!

**Comment by CadiChris** [comment]
Hi @salvoravida . Thanks for the update!

I won't be able to give you immediate feedback, as updating the authentication library version in my product right now is not a move I'll make, I'm sorry.

I'm not really confortable with updating the core `adal` version in my current environment.

Based on the code changes I've read, I don't see why it would not work as expected 👍
